contracts/core/MarginRequirements.sol

```diff
@@ -0,0 +1,223 @@
+        address _collateralAsset,
+        bool _isPut,
+        address _account,
+        uint256 _initialMargin
+    ) external onlyOwner {
+        require(
+            _initialMargin > 0 && _initialMargin <= 100 * 10**2,
+            "MarginRequirements: initial margin cannot be 0 or higher than 100%"
+        );
+        require(_underlying != address(0), "MarginRequirements: invalid underlying");
+        require(_collateralAsset != address(0), "MarginRequirements: invalid collateral");
+        require(_account != address(0), "MarginRequirements: invalid account");
+
+        initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][_account] = _initialMargin;
+    }
+
+    /**
+     * @notice sets the maintenance margin absolute amount
+     * @dev can only be called by keeper
+     * @param _account account address
+     * @param _vaultID id of the vault
+     * @param _maintenanceMargin maintenance margin absolute amount
+     */
+    function setMaintenanceMargin(
+        address _account,
+        uint256 _vaultID,
+        uint256 _maintenanceMargin
+    ) external onlyKeeper {
+        require(_maintenanceMargin > 0, "MarginRequirements: initial margin cannot be 0");
+        require(_account != address(0), "MarginRequirements: invalid account");
+
+        maintenanceMargin[_account][_vaultID] = _maintenanceMargin;
+    }
+
+    /************************************************
+     *  MARGIN OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice clears the maintenance margin mapping
+     * @dev can only be called by OTC wrapper contract
+     * @param _account account address
+     * @param _vaultID id of the vault
+     */
+    function clearMaintenanceMargin(address _account, uint256 _vaultID) external onlyOTCWrapper {
+        delete maintenanceMargin[_account][_vaultID];
+    }
+
+    /**
+     * @notice checks if there is enough collateral to mint the desired amount of otokens
+     * @param _account account address
+     * @param _notional order notional amount
+     * @param _underlying underlying asset address
+     * @param _isPut option type the vault is selling
+     * @param _collateralAsset collateral asset address
+     * @param _collateralAmount collateral amount
+     * @return boolean value stating whether there is enough collateral to mint
+     */
+    function checkMintCollateral(
+        address _account,
+        uint256 _notional,
+        address _underlying,
+        bool _isPut,
+        uint256 _collateralAmount,
+        address _collateralAsset
+    ) external view returns (bool) {
+        uint256 collateralDecimals = uint256(ERC20Interface(_collateralAsset).decimals());
+
+        uint256 initialMarginRequired = initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][
+            _account
+        ];
+
+        return
+            _notional.mul(initialMarginRequired).mul(10**collateralDecimals).mul(1e8) <
```

Comment:
can you elaborate on the `.mul(10**collateralDecimals).mul(1e8)` and `.mul(100e2).mul(1e6)`

notional is 6 decimals, initial margin is 2 decimals. 

_collateralAmount is decimals of the asset, oracle.getPrice(_collateralAsset) is 8 decimals,

Reply:
By item:

- notional = x * 10^6
- initialMarginRequired = [ y  / (100 * 10^2) ] 
- _vault.collateralAmounts[0].sub(_withdrawAmount).sub(maintenanceMargin[_vaultID]) = z * 10^(collateralDecimals)
- oracle.getPrice(_vault.collateralAssets[0]) = a * 10^8

1) x * [ y / (100 * 10^2) ] < z * a -> if we removed decimals from all the numbers this would be what we want

2) Then I add back the decimals to both sides of the equation  to make sure both sides have the same amount of decimals
- x: 10^6 gets mirrored on the other side of the equation with a `1e6`

`notional` * [ y / (100 * 10^2) ] < z * a * `1e6`

- a: 10^8 gets mirrored on the other side of the equation with a `1e8`

`notional` * [ y / (100 * 10^2) ] * `1e8` < z * `oracle.getPrice(_vault.collateralAssets[0])` * `1e6`

- z: 10^(collateralDecimals) gets mirrored on the other side of the equation with a `10**collateralDecimals`


`notional` * [ y / (100 * 10^2) ] * `1e8` * `10**collateralDecimals` < `_vault.collateralAmounts[0].sub(_withdrawAmount).sub(maintenanceMargin[_vaultID])`  * `oracle.getPrice(_vault.collateralAssets[0])` * `1e6`

- y: (100 * 10^2) is dividing on the left side of the equation so it's moved as multiplication to the right side `100e2`. Since the result of the division is zero decimals there is no need to mirror anything on the other side of the equation.

`notional` *  `y`  * `1e8` * `10**collateralDecimals` < `_vault.collateralAmounts[0].sub(_withdrawAmount).sub(maintenanceMargin[_vaultID])`  * `oracle.getPrice(_vault.collateralAssets[0])` * `1e6` * `100e2`

In this way, both sides of the equation end up with the same amount of decimals and there is no need for divisions.